File: C#/0599. Minimum Index Sum of Two Lists.cs
public class Solution {
    public string[] FindRestaurant(string[] list1, string[] list2) {
        Dictionary<string,int> list1Dict=new Dictionary<string,int>();
        for(int i=0;i<list1.Count();i++){
            list1Dict[list1[i]]=i;
        }
        IList<string> rep=new List<string>();
        int totalInterest=int.MaxValue;
        
        for(int i=0;i<list2.Count();i++){
            if(list1Dict.ContainsKey(list2[i])){
                if(list1Dict[list2[i]]+i==totalInterest){
                    rep.Add(list2[i]);
                }
                else if(list1Dict[list2[i]]+i<totalInterest){
                    rep.Clear();
                    rep.Add(list2[i]);
                    totalInterest=list1Dict[list2[i]]+i;
                }
            }    
        }
        
        return rep.ToArray();
    }
}

File: 20190519python/0599. Minimum Index Sum of Two Lists.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        dict_list={}
        for k in range(len(list1)):
            dict_list[list1[k]]=k
        
        rep=[]
        cnt=len(list1)+len(list2)
        
        for k in range(len(list2)):
            if list2[k] in dict_list:
                if k+dict_list[list2[k]]<cnt:
                    cnt=k+dict_list[list2[k]]
                    rep=[list2[k]]
                elif k+dict_list[list2[k]]==cnt:
                    rep.append(list2[k])
        
        return rep
